### PR TITLE
Create testlang for language independent tests

### PR DESCRIPTION
**Pull request by AntonyBlakey**
This creates a new language `testlang` that is used purely for testing. The tests that are not absolutely solidity specific are moved to the testlang subtree.

### CODE REVIEW

File: Cargo.toml
Comment (OmarTawfik) [inline Cargo.toml:23]: nit: I suggest keeping the `members` entries sorted alphabetically.

File: Cargo.toml
Comment (OmarTawfik) [inline Cargo.toml:60]: I think this might need coordination with the changes from #745, whichever lands sooner. cc @Xanewok

File: crates/infra/cli/src/toolchains/napi/resolver.rs
Comment (OmarTawfik) [inline crates/infra/cli/src/toolchains/napi/resolver.rs:13]: I suggest using an `enum` for the name here (`NapiResolver::new(NapiPackageId::Solidity)`), or exposing two constructors instead (`NapiResolver::solidity()` and `NapiResolver::test_lang()`).

- The path will be stored in a single place (here in the resolver) to make it easier to maintain the current infra scripts.
- This will make sure that callsite is type checked, which also makes it easier to refactor out and make these build scripts generic later.
- It lets us use complete package names in string literals, instead of `format!()`, which allows us to do string search/replace over the whole codebase.

Comment (AntonyBlakey) [reply crates/infra/cli/src/toolchains/napi/resolver.rs:13]: The reason I use a string value is to make external configuration possible, rather than encoding the package structure in the type system. However I acknowledge that parametricity is pointless in this case.

I'll add constructors.

File: Cargo.toml
Comment (AntonyBlakey) [reply Cargo.toml:60]: Done

File: Cargo.toml
Comment (AntonyBlakey) [reply Cargo.toml:23]: This was because testlang is more 'core' than solidity, but done.

File: crates/codegen/parser/runtime/src/templates/language.rs.jinja2
Comment (OmarTawfik) [inline crates/codegen/parser/runtime/src/templates/language.rs.jinja2:5]: nit: sorting.

Comment (AntonyBlakey) [reply crates/codegen/parser/runtime/src/templates/language.rs.jinja2:5]: Done

File: crates/infra/cli/generated/infra.zsh-completions
Comment (OmarTawfik) [inline crates/infra/cli/generated/infra.zsh-completions:13]: I think file is left over from rebase?

Comment (AntonyBlakey) [reply crates/infra/cli/generated/infra.zsh-completions:13]: Done

File: crates/infra/cli/src/toolchains/napi/resolver.rs
Comment (OmarTawfik) [reply crates/infra/cli/src/toolchains/napi/resolver.rs:13]: Sounds great. Thank you! In that case, and for the third reason above, I suggest storing the full package names. I realize the existing code indicates that they should share the same prefix, but using full string literals for package/folder identifiers is really important to keep infra maintainable:

```rust
    pub fn testlang() -> Self {
        Self {
            rust_crate: "testlang_npm_crate",
            typescript_crate: "testlang_npm_package",
        }
    }

    pub fn solidity() -> Self {
        Self {
            rust_crate: "solidity_npm_crate",
            typescript_crate: "solidity_npm_package",
        }
    }
```

Comment (AntonyBlakey) [reply crates/infra/cli/src/toolchains/napi/resolver.rs:13]: Done

File: crates/testlang/outputs/cargo/tests/src/versions/mod.rs
Comment (OmarTawfik) [inline crates/testlang/outputs/cargo/tests/src/versions/mod.rs:4]: I wonder why these tests were moved out of `Solidity`?
For example, this one in perticular asserts the minimum solidity supported version, which is a user-facing piece of information.

Comment (OmarTawfik) [reply crates/testlang/outputs/cargo/tests/src/versions/mod.rs:4]: If the intent is for the tests in `testlang` to expand to test other Slang-properties (not specific to Solidity), then I suggest (at least for now) to just copy the tests instead of moving them, and they can evolve separately from there.

File: crates/testlang/outputs/cargo/tests/src/lib.rs
Comment (OmarTawfik) [inline crates/testlang/outputs/cargo/tests/src/lib.rs:3]: I wonder if these imports are needed? is it possible to just remove the deps from the new crate?

File: crates/testlang/outputs/cargo/tests/Cargo.toml
Comment (OmarTawfik) [inline crates/testlang/outputs/cargo/tests/Cargo.toml:8]: I wonder if `[build-dependencies]` is needed here as well, given there is no `build.rs`?

File: crates/testlang/outputs/cargo/tests/src/versions/mod.rs
Comment (AntonyBlakey) [reply crates/testlang/outputs/cargo/tests/src/versions/mod.rs:4]: Because IMO they add no real value that is solidity specific. They have no real coverage, and if they did they would merely replicate the data. They merely test the versions mechanism, which is not solidity specific.

To test the minimum version you would need to test every possible version before 0.4.11 to verify that it is the minimum. Otherwise it is only testing that 0.4.11 is included, which is in the data.

File: crates/testlang/outputs/cargo/tests/Cargo.toml
Comment (AntonyBlakey) [reply crates/testlang/outputs/cargo/tests/Cargo.toml:8]: Done

File: crates/testlang/outputs/cargo/tests/src/lib.rs
Comment (AntonyBlakey) [reply crates/testlang/outputs/cargo/tests/src/lib.rs:3]: Done

File: crates/infra/cli/src/toolchains/napi/compiler.rs
Comment (OmarTawfik) [inline crates/infra/cli/src/toolchains/napi/compiler.rs:54]: Not sure if we need to add/maintain the packages in `crates/testlang/outputs/npm/package/platforms`? since they are only used for release. If not, I suggest:

1. Removing the platform-specific packages from `crates/testlang/outputs/npm/package/platforms`.
2. Clearing `additional[]` in `crates/testlang/outputs/npm/package/package.json`.
3. Adding a check to `compiler.rs` for this. e.g.:

```suggestion
    let targets = NapiConfig::list_all_targets(resolver)?;

    assert_ne!(targets.len(), 0, "No release targets found. Is this a publicly released package?");
```

File: crates/testlang/outputs/npm/package/package.json
Comment (OmarTawfik) [inline crates/testlang/outputs/npm/package/package.json:67]: To avoid publishing this package, I suggest removing `publishConfig` here, and adding `private: true` above.

```suggestion
```

File: crates/testlang/outputs/npm/package/package.json
Comment (OmarTawfik) [inline crates/testlang/outputs/npm/package/package.json:3]: To avoid publishing this package, I suggest removing the `publishConfig` entry, and adding `private: true`:

```suggestion
  "version": "0.13.1",
  "private": true,
```

File: crates/testlang/outputs/npm/package/package.json
Comment (OmarTawfik) [inline crates/testlang/outputs/npm/package/package.json:6]: I also suggest removing all the additional metadata, as they are only used for public facing artifacts (across languages), and are maintained together: description, URLS, contributors, keywords, etc...

File: crates/solidity/outputs/npm/tests/src/tests/ast.ts
Comment (OmarTawfik) [inline crates/solidity/outputs/npm/tests/src/tests/ast.ts:1]: Same suggestion for these tests, copying/forking the tests rather than moving.

I'm a bit uncomfortable about removing such E2E tests for Solidity and its types. This is the only coverage it has.

File: crates/testlang/outputs/spec/Cargo.toml
Comment (OmarTawfik) [inline crates/testlang/outputs/spec/Cargo.toml:3]: I understand the need for Rust/TypeScript outputs for `testlang`, but I wonder if we need the spec, given that it is a relatively heavy step, with lots of files to add/maintain manually over time.
Are there any consumers for this? if not, I suggest:

- removing `testlang/outputs/spec` crate.
- removing `testlang/inputs/language/docs` dir.

File: crates/testlang/inputs/language/src/definition.rs
Comment (OmarTawfik) [inline crates/testlang/inputs/language/src/definition.rs:222]: nit: left over?

File: crates/testlang/outputs/cargo/crate/Cargo.toml
Comment (OmarTawfik) [inline crates/testlang/outputs/cargo/crate/Cargo.toml:2]: I also suggest removing all the additional metadata, as they are only used for public facing artifacts (across languages), and are maintained together: description, URLS, contributors, keywords, etc...

File: crates/testlang/outputs/cargo/crate/Cargo.toml
Comment (OmarTawfik) [inline crates/testlang/outputs/cargo/crate/Cargo.toml:5]: I suggest removing the `publish = true`, so that it is ingored in CD.

File: crates/testlang/outputs/cargo/crate/src/main.rs
Comment (OmarTawfik) [inline crates/testlang/outputs/cargo/crate/src/main.rs:27]: nit: left over `.sol` below.

File: crates/testlang/outputs/cargo/crate/src/main.rs
Comment (OmarTawfik) [inline crates/testlang/outputs/cargo/crate/src/main.rs:20]: Are we planning on consuming this binary?
If not, I suggest removing `main.rs`, along with the `mod supress` in `lib.rs` as it will be no longer needed.

File: crates/testlang/outputs/spec/generated/public/supported-versions.md
Comment (OmarTawfik) [inline crates/testlang/outputs/spec/generated/public/supported-versions.md:3]: Found a bug in `codegen_spec`!
I will submit a fix for it after this PR though. No reason to block on it.

```suggestion
This specification compiles information from 4 publicly released versions of Testlang:
```

File: crates/testlang/testing/utils/Cargo.toml
Comment (OmarTawfik) [inline crates/testlang/testing/utils/Cargo.toml:2]: is `testlang_testing_utils` used anywhere? If not I suggest removing it.
The matching Solidity APIs are only there for pragma extraction, and should be removed alltogether in #806

**Review by OmarTawfik — CHANGES_REQUESTED** [review]
Left a few suggestions/concerns

File: crates/testlang/outputs/cargo/tests/src/versions/mod.rs
Comment (OmarTawfik) [reply crates/testlang/outputs/cargo/tests/src/versions/mod.rs:4]: I agree in principle. The tests make sure we do generate a valid sorted `SUPPORTED_VERSIONS`, which can be tested in `testlang`. But practically speaking, given that (excluding CST snapshots), we have less than 15 unit tests covering the entire public API, which barely qualifies as a "sanity check". Given how little effort it takes to maintain them (so far), I think it is worth it to keep them.

While we have some degree of confidence in our build/codegen pipeline that it will maintain both languages in check, accidental divergence and bugs can and will happen (I just discovered a tiny one while reviewing this PR: #812), and the tests will prevent "cross contamination".

File: crates/testlang/outputs/spec/Cargo.toml
Comment (OmarTawfik) [reply crates/testlang/outputs/spec/Cargo.toml:3]: On the other hand, because of this, I discovered #812 ...

File: crates/testlang/outputs/npm/package/package.json
Comment (AntonyBlakey) [reply crates/testlang/outputs/npm/package/package.json:3]: Done

File: crates/testlang/outputs/npm/package/package.json
Comment (AntonyBlakey) [reply crates/testlang/outputs/npm/package/package.json:6]: Done

File: crates/testlang/outputs/cargo/tests/src/versions/mod.rs
Comment (AntonyBlakey) [reply crates/testlang/outputs/cargo/tests/src/versions/mod.rs:4]: Duplicated tests are the same as one test. So keeping the tests in the solidity crates/package adds nothing to test coverage or test value. IMO we should eliminate the reduncdancy that adds no actual value. To think that it adds to our test coverage is wrong, and serves merely to mislead us.

And I strongly feel that testing in the solidity crate should be about the grammar, and other things that are entirely specific to the language. Everything else about slang should be tested in the non-solidity crates/packages.

Relying on some random test duplication to detect possible divergence in the tiny part of overlap is also misleading security.

Lets F2F this.

File: crates/testlang/outputs/spec/Cargo.toml
Comment (AntonyBlakey) [reply crates/testlang/outputs/spec/Cargo.toml:3]: Given that we don't test the doc generation in any way, we should remove it.

File: crates/testlang/outputs/cargo/crate/Cargo.toml
Comment (AntonyBlakey) [reply crates/testlang/outputs/cargo/crate/Cargo.toml:5]: Done

File: crates/solidity/outputs/npm/tests/src/tests/ast.ts
Comment (AntonyBlakey) [reply crates/solidity/outputs/npm/tests/src/tests/ast.ts:1]: As per my other comment on this topic

File: crates/testlang/inputs/language/src/definition.rs
Comment (AntonyBlakey) [reply crates/testlang/inputs/language/src/definition.rs:222]: Done

File: crates/infra/cli/src/toolchains/napi/compiler.rs
Comment (AntonyBlakey) [reply crates/infra/cli/src/toolchains/napi/compiler.rs:54]: Done

File: crates/testlang/outputs/cargo/crate/src/main.rs
Comment (AntonyBlakey) [reply crates/testlang/outputs/cargo/crate/src/main.rs]: Done

File: crates/testlang/outputs/cargo/crate/src/main.rs
Comment (AntonyBlakey) [reply crates/testlang/outputs/cargo/crate/src/main.rs]: Done

File: crates/testlang/outputs/spec/Cargo.toml
Comment (OmarTawfik) [reply crates/testlang/outputs/spec/Cargo.toml]: Sounds good.

File: crates/infra/cli/src/toolchains/napi/compiler.rs
Comment (OmarTawfik) [inline crates/infra/cli/src/toolchains/napi/compiler.rs:133]: I'm a bit worried about this condition, as it allowing shipping a release without a license or other critical files. I think we should fail the build if not found.

To avoid unnecissarily complicating this module logic, what do you think of adding dummy/empty versions of these files under `testlang`?

Comment (OmarTawfik) [inline crates/infra/cli/src/toolchains/napi/compiler.rs:133]: I'm a bit worried about this condition, as it allowing shipping a release without a license or other critical files (which can accidentally happen as we move stuff around). I think we should fail the build if not found.

To avoid further adding to this module logic, what do you think of adding dummy/empty versions of these files under `testlang`?

File: crates/testlang/outputs/cargo/slang_testlang/build.rs
Comment (OmarTawfik) [inline crates/testlang/outputs/cargo/slang_testlang/build.rs:4]: All 3 `clippy::cargo_common_metadata` are not actually triggering for me. Not sure if these are needed?

Comment (AntonyBlakey) [reply crates/testlang/outputs/cargo/slang_testlang/build.rs:4]: It was failing for me, not sure wy, but now it is not.

File: crates/testlang/outputs/cargo/slang_testlang/build.rs
Comment (OmarTawfik) [inline crates/testlang/outputs/cargo/slang_testlang/build.rs:1]: nit: lingering comment

**Review by OmarTawfik — APPROVED** [review]
Looks great to me! Thank you.

I resolved all but two last comments (nit clean up before merging). Accepted to unblock you.

File: crates/testlang/outputs/cargo/slang_testlang/build.rs
Comment (AntonyBlakey) [reply crates/testlang/outputs/cargo/slang_testlang/build.rs:1]: Done

File: crates/testlang/outputs/cargo/slang_testlang/build.rs
Comment (AntonyBlakey) [reply crates/testlang/outputs/cargo/slang_testlang/build.rs]: Done

**Pull request merged**
